credentials/tls/certprovider/meshca/plugin.go

```diff
@@ -0,0 +1,280 @@
+	meshpb "google.golang.org/grpc/credentials/tls/certprovider/meshca/internal/v1"
+	"google.golang.org/grpc/internal/grpclog"
+	"google.golang.org/grpc/metadata"
+)
+
+// In requests sent to the MeshCA, we add a metadata header with this key and
+// the value being the GCE zone in which the workload is running in.
+const locationMetadataKey = "x-goog-request-params"
+
+// providerPlugin is an implementation of the certprovider.Provider interface,
+// which gets certificates signed by communicating with the MeshCA.
+type providerPlugin struct {
+	cancel   context.CancelFunc
+	cc       *grpc.ClientConn          // Connection to MeshCA server.
+	cfg      *pluginConfig             // Plugin configuration.
+	opts     certprovider.Options      // Key material options.
+	dist     *certprovider.Distributor // Holds the key material.
+	logger   *grpclog.PrefixLogger     // Plugin instance specific prefix.
+	backoff  func(int) time.Duration   // Exponential backoff.
+	doneFunc func()                    // Notify the builder when done.
+}
+
+// providerParams wraps params passed to the provider plugin at creation time.
+type providerParams struct {
+	// This ClientConn to the MeshCA server is owned by the builder.
+	cc       *grpc.ClientConn
+	cfg      *pluginConfig
+	opts     certprovider.Options
+	backoff  func(int) time.Duration
+	doneFunc func()
+}
+
+func newProviderPlugin(params providerParams) *providerPlugin {
+	ctx, cancel := context.WithCancel(context.Background())
+	p := &providerPlugin{
+		cancel:   cancel,
+		cc:       params.cc,
+		cfg:      params.cfg,
+		opts:     params.opts,
+		backoff:  params.backoff,
+		doneFunc: params.doneFunc,
+		dist:     certprovider.NewDistributor(),
+	}
+	p.logger = prefixLogger((p))
+	p.logger.Infof("plugin created")
+	go p.run(ctx)
+	return p
+}
+
+func (p *providerPlugin) KeyMaterial(ctx context.Context) (*certprovider.KeyMaterial, error) {
+	return p.dist.KeyMaterial(ctx)
+}
+
+func (p *providerPlugin) Close() {
+	p.logger.Infof("plugin closed")
+	p.dist.Stop()
+	p.cancel()
+	p.doneFunc()
+}
+
+// run is a long running goroutine which periodically sends out CSRs to the
+// MeshCA, and updates the underlying Distributor with the new key material.
+func (p *providerPlugin) run(ctx context.Context) {
+	// We need to start fetching key material right away. The next attempt will
+	// be triggered by the timer firing.
+	certValidity := p.updateKeyMaterial(ctx)
+	if certValidity == 0 {
+		return
+	}
+
+	for {
+		// We request a certificate whose validity is usually twice as much as
+		// the grace period. But the server is free to return a certificate with
+		// whatever validity time it deems right.
+		refreshAfter := p.cfg.certGraceTime
+		if refreshAfter > certValidity {
+			refreshAfter = certValidity
+		}
+		timer := time.NewTimer(refreshAfter)
+		select {
+		case <-ctx.Done():
+			timer.Stop()
+			return
+		case <-timer.C:
+			certValidity = p.updateKeyMaterial(ctx)
```

Comment:
So if this is zero, like in the initial attempt, what happens?  We seem to attempt to poll immediately forever?

Reply:
Aha.  So maybe rework this function so `updateKeyMaterial` is only called once, at the top of the for loop.